pyspider/libs/base_queue.py

```diff
@@ -0,0 +1,101 @@
+    first process. The solution is to use a multiprocessing.Lock to guarantee
+    the atomicity of the modifications to Value.
+    This class comes almost entirely from Eli Bendersky's blog:
+    http://eli.thegreenplace.net/2012/01/04/shared-counter-with-pythons-multiprocessing/
+    """
+
+    def __init__(self, n=0):
+        self.count = multiprocessing.Value('i', n)
+
+    def increment(self, n=1):
+        """ Increment the counter by n (default = 1) """
+        with self.count.get_lock():
+            self.count.value += n
+
+    @property
+    def value(self):
+        """ Return the value of the counter """
+        return self.count.value
+
+
+class Queue(BaseQueue.Queue, object):
+    """ A portable implementation of multiprocessing.Queue.
+    Because of multithreading / multiprocessing semantics, Queue.qsize() may
+    raise the NotImplementedError exception on Unix platforms like Mac OS X
+    where sem_getvalue() is not implemented. This subclass addresses this
+    problem by using a synchronized shared counter (initialized to zero) and
+    increasing / decreasing its value every time the put() and get() methods
+    are called, respectively. This not only prevents NotImplementedError from
+    being raised, but also allows us to implement a reliable version of both
+    qsize() and empty().
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(Queue, self).__init__(*args, **kwargs)
+        self.size = SharedCounter(0)
+
+    def put(self, *args, **kwargs):
+        self.size.increment(1)
+        super(Queue, self).put(*args, **kwargs)
+
+    def get(self, *args, **kwargs):
+        v = super(Queue, self).get(*args, **kwargs)
+        self.size.increment(-1)
+        return v
+
+    def qsize(self):
+        """ Reliable implementation of multiprocessing.Queue.qsize() """
+        return self.size.value
+
+    def empty(self):
+        """ Reliable implementation of multiprocessing.Queue.empty() """
+        return not self.qsize()
+
+
+class MultiProcessingQueue(MPQueue, object):
+    def __init__(self, *args, **kwargs):
+        super(MultiProcessingQueue, self).__init__(*args, **kwargs)
+        self.size = SharedCounter(0)
+
+    def put(self, *args, **kwargs):
+        self.size.increment(1)
+        super(MultiProcessingQueue, self).put(*args, **kwargs)
+
+    def get(self, *args, **kwargs):
+        v = super(MultiProcessingQueue, self).get(*args, **kwargs)
+        self.size.increment(-1)
+        return v
+
+    def qsize(self):
+        """ Reliable implementation of multiprocessing.Queue.qsize() """
+        return self.size.value
+
+    def empty(self):
+        """ Reliable implementation of multiprocessing.Queue.empty() """
+        return not self.qsize()
+
+
+def get_multiprocessing_queue(maxsize=0):
+    if hasattr(multiprocessing, 'get_context'):  # python 3.4
+        return MultiProcessingQueue(maxsize,
+                                    ctx=multiprocessing.get_context())
```

Comment:
[Queue.__init__](https://github.com/python/cpython/blob/master/Lib/multiprocessing/queues.py#L36) is defined like this:

    def __init__(self, maxsize=0, *, ctx):

And [multiprocessing/context.py](https://github.com/python/cpython/blob/master/Lib/multiprocessing/context.py#L98) have some magic hide the detail that make you feel like ctx is not exists:

    def Queue(self, maxsize=0):
        '''Returns a queue object'''
        from .queues import Queue
        return Queue(maxsize, ctx=self.get_context())

But, you must get your hands dirty if you override Queue class (and others).